mim/utils/mmpack/pack_cfg.py

```diff
@@ -0,0 +1,302 @@
+
+    cfg.dump(cfg_pth)
+
+    _replace_config_scope_to_pack(cfg_pth)
+
+    # transform to default_scope
+    init_default_scope(default_scope)
+
+    # wrap ``Registry.build()`` for exporting modules
+    _wrapper_all_registries_build_func(
+        export_module_dir=export_module_dir, scope=default_scope)
+
+    cfg['work_dir'] = osp.join(export_root_dir,
+                               'export_log')  # creat temp work_dirs for export
+
+    def error_postprocess():
+        if osp.exists(export_root_dir):
+            shutil.rmtree(export_root_dir)
+        origin_cfg.dump(cfg_pth.split('/')[-1])
+        traceback.print_exc()
+
+    # use runner to export all needed modules
+    runner = Runner.from_cfg(cfg)
+    try:
+        runner.build_train_loop(cfg.train_cfg)
+    except FileNotFoundError:
+        error_postprocess()
+        print(f"[\033[94m Debug \033[0m] The data root of 'train_dataloader' "
+              f"is not found. Please modify the 'data_root' in "
+              f"duplicate config '\033[1m{cfg_pth.split('/')[-1]}\033[0m'.")
+        return -1
+
+    try:
+        if 'val_cfg' in cfg and cfg.val_cfg is not None:
+            runner.build_val_loop(cfg.val_cfg)
+    except FileNotFoundError:
+        error_postprocess()
+        print(f"[\033[94m Debug \033[0m] The data root of 'val_dataloader' "
+              f"is not found. Please modify the 'data_root' in duplicate "
+              f"config '\033[1m{cfg_pth.split('/')[-1]}\033[0m'.")
+        return -1
+
+    try:
+        if 'test_cfg' in cfg and cfg.test_cfg is not None:
+            runner.build_test_loop(cfg.test_cfg)
+    except FileNotFoundError:
+        error_postprocess()
+        print(f"[\033[94m Debug \033[0m] The data root of 'test_dataloader' "
+              f"is not found. Please modify the 'data_root' in duplicate "
+              f"config '\033[1m{cfg_pth.split('/')[-1]}\033[0m'.")
+        return -1
+
+    if 'optim_wrapper' in cfg and cfg.optim_wrapper is not None:
+        runner.optim_wrapper = runner.build_optim_wrapper(cfg.optim_wrapper)
+    if 'param_scheduler' in cfg and cfg.param_scheduler is not None:
+        runner.build_param_scheduler(cfg.param_scheduler)
+
+    # add ``__init__.py`` to all dirs, for transferring directories
+    # to be modules
+    for directory, _, _ in os.walk(export_module_dir):
+        if not osp.exists(osp.join(directory, '__init__.py')) \
+                and 'configs' not in directory \
+                and not directory.endswith(f'{export_module_dir}/'):
+            with open(osp.join(directory, '__init__.py'), 'w') as f:
+                f.write(__init__str)
+
+    # get tools from web
+    tools_dir = osp.join(export_root_dir, 'tools')
+    os.makedirs(tools_dir, exist_ok=True)
+
+    # postprocess for ``pack/registry.py``
+    with open(
+            osp.join(export_module_dir, 'registry.py'), encoding='utf-8') as f:
+        ast_tree = ast.parse(f.read())
+
+    for node in ast.walk(ast_tree):
```

Comment:
Suggest encapsulating all AST-related code snippets into a function, keeping this interface focused on high-level logic